rfcs/20220217-visual_regression_testing.md

```diff
@@ -0,0 +1,240 @@
+
+Percy definitely looks very cool and I would recommend creating a BrowserStack account and checking out the Percy demo project. It's fun to poke around in and the UI mostly works very nicely.
+
+However, if I'm being honest, when taking into account the fact that GitHub already has image diffing built into the PR review process, Percy seems like an unnecessary additional complication to the baseline of visual regression testing. Maybe it's something that we would come to realize the value of down the line, but it does not seem to me to be a requirement for the initial phase of visual regression testing.
+
+#### Alternative image snapshot libraries
+
+Playwrights's built in image snapshot-diffing capability works fine for the most part. In fact, it basically does everything you need short of one very useful feature: it throws away the actual generated diff.
+
+[Amex's `jest-image-snapshot`](https://github.com/americanexpress/jest-image-snapshot) fixes this. In addition to failing when snapshots do not match, it will also write the diff image that includes the "hot spots" where the visual differences exist. It also automatically generates the snapshot name, which Playwright's matcher doesn't seem to be capable of (this is usually a standard feature of Jest snapshot matchers otherwise).
+
+`jest-image-snapshot` is also significantly more configurable than Playwright's matcher. Playwright's matcher has a single option, `threshold` which it passes directly to `pixelmatch`. Amex's library has a [ton of potentially useful options for configuring snapshots on an individual basis or project wide](https://github.com/americanexpress/jest-image-snapshot#%EF%B8%8F-api) and also still uses `pixelmatch` in the background anyway.
+
+**Important caveat to this section!**
+
+I ended up trying out `jest-image-snapshot` in https://github.com/WordPress/openverse-frontend/pull/899 but running the matcher fails on some error about `_counters` being undefined. It looks like Playwright, while it uses the jest `expect` library, does not support the same matcher context as Jest proper does. To use `jest-image-snapshot` we might have to switch to `jest-playwright` which would be less than ideal as it's [recommended against by the Playwright project](https://playwright.dev/docs/test-runners#jest--jasmine) (though for what reason exactly I'm really not sure).
+
+We can still _try_ to find a workaround for this, but if it ends up being too much trouble I'd say just leave it out of the final roadmap for this feature and we can use the built-in Playwright tool. Playwright will output the before and after images for failed snapshot comparisons into a `test-results` folder at the top level of the project, so I figure the worst case scenario is that we could write a little script to walk through that directory and run `pixelmatch` on them to get the diff image.
+
+## Repository size considerations
```

Comment:
We can also use an extension like [git-lfs](https://git-lfs.github.com/) (Git Large File Storage) to keep the images entirely separate from the actual codebase but still part of the same repository. Git LFS can be a pain to work with, but I believe you can work on the codebase without having to interact with it (e.g. download the files) at all.

Reply:
You know, that crossed my mind but I don't know why I didn't pursue that idea further. It seems like exactly the right solution for this problem (if indeed it turns out to be a problem cause like I said the snapshot images are pretty tiny)

I have also heard that git lfs can be a pain, but it's gotten better in the past couple years as I understand it.